### PR TITLE
Allow fractional answers in `pl-matrix-component-input`

### LINKED ISSUE #2235: Allow fractional inputs in `pl-matrix-component-input`
Math415 students were very happy with the new fractional answers in `pl-number-input`, and allowing fractional answers in the matrix input will make them even happier.

### PR DESCRIPTION
**Pull request by nicknytko**
Adds an attribute `allow-fractions`, default `True`, to the matrix component input element.

Fixes #2235

### CODE REVIEW

**Comment by coatless** [comment]
Errors with:

```
         the test job sequence
           should be successful:

      AssertionError: expected 'Error' to equal 'Success'
      + expected - actual

      -Error
      +Success
      
      at Context.<anonymous> (tests/helperQuestion.js:666:28)
      at processImmediate (internal/timers.js:456:21)
      at process.topLevelDomainCallback (domain.js:137:15)

  2) Auto-test questions in exampleCourse
       auto-testing question elementMatrixComponentInput
         the test job sequence
           should produce no issues:
     found 2 issues (expected zero issues):
```

**Comment by nicknytko** [comment]
@coatless thanks for digging through the test output!  I just haven't had a chance recently to fix this up.

File: elements/pl-matrix-component-input/pl-matrix-component-input.py [hunk 1]
```diff
@@ -210,9 +213,77 @@ def render(element_html, data):
     return html
 
 
+def parse_entry(name, data, allow_fractions):
```
Comment (coatless) [inline elements/pl-matrix-component-input/pl-matrix-component-input.py]: Maybe this should be embedded in `pl` module as it is now shared across multiple elements?

File: elements/pl-matrix-component-input/pl-matrix-component-input.py [hunk 2]
```diff
@@ -232,32 +303,15 @@ def parse(element_html, data):
...
-            a_sub = data['submitted_answers'].get(each_entry_name, None)
-            if a_sub is None:
-                data['submitted_answers'][each_entry_name] = None
-                data['format_errors'][each_entry_name] = '(No submitted answer)'
-                invalid_format = True
-            elif not a_sub:
-                data['submitted_answers'][each_entry_name] = None
-                data['format_errors'][each_entry_name] = '(Invalid blank entry)'
-                invalid_format = True
+            value, valid = parse_entry(each_entry_name, data, allow_fractions)
```
Comment (coatless) [inline elements/pl-matrix-component-input/pl-matrix-component-input.py]: Very nice simplification.

File: elements/pl-matrix-component-input/pl-matrix-component-input.py (hunk 1 above)
Comment (nicknytko) [reply elements/pl-matrix-component-input/pl-matrix-component-input.py:216]: Yeah, after looking at the other functions in there I think that would be best in terms of not repeating code

File: exampleCourse/questions/elementMatrixComponentInput/server.py [hunk 3]
```diff
@@ -11,7 +11,8 @@ def generate(data):
     sf = 2
     B = np.round(A,sf)
     x =  np.array([[1,2,3,4]])
-    long_matrix = np.array([[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]])
+    long_matrix = np.arange(1, 21).reshape((1, 20))
```
Comment (coatless) [inline exampleCourse/questions/elementMatrixComponentInput/server.py:14]: Huzzah! Nice use of `.arange()`

File: exampleCourse/questions/elementMatrixComponentInput/question.html [hunk 4]
```diff
@@ -70,9 +70,24 @@ <h5>
...
 </div>
+
+<div class="mb-2" >
+    <h5>
+        5.
+    </h5>
+    <pl-question-panel>
+        <p>
+            Enter the vector
+            $${\bf y} = \begin{bmatrix} \frac{1}{2} & \frac{1}{4} & \frac{1}{8} & \frac{1}{16} \end{bmatrix}$$
```
Comment (coatless) [inline exampleCourse/questions/elementMatrixComponentInput/question.html:87]: Thanks! This is great.

File: elements/pl-number-input/pl-number-input.py [hunk 5]
```diff
@@ -245,62 +245,14 @@ def parse(element_html, data):
...
-    if a_sub is None:
-        data['format_errors'][name] = 'No submitted answer.'
-        data['submitted_answers'][name] = None
-        return
-
-    if a_sub.strip() == '':
-        data['format_errors'][name] = get_format_string(allow_complex, allow_fractions, 'the submitted answer was blank.')
-        data['submitted_answers'][name] = None
-        return
+    value, newdata = pl.string_fraction_to_number(a_sub, allow_fractions, allow_complex)
```
Comment (coatless) [inline elements/pl-number-input/pl-number-input.py:249]: Wow this really simplified things in `pl-number-input`.

File: question-servers/freeformPythonLib/prairielearn.py [hunk 6]
```diff
@@ -631,6 +631,80 @@ def string_to_number(s, allow_complex=True):
...
+        # Specially handle fractions.
+
+        if allow_fractions:
+            a_sub_splt = a_sub.split('/')
+            try:
+                a_parse_l = string_to_number(a_sub_splt[0], allow_complex=allow_complex)
+                a_parse_r = string_to_number(a_sub_splt[1], allow_complex=allow_complex)
+
+                if a_parse_l is None or not np.isfinite(a_parse_l):
+                    raise ValueError(f'The numerator could not be interpreted as a decimal{ or_complex }number.')
```
Comment (coatless) [inline question-servers/freeformPythonLib/prairielearn.py:676]: Nice the errors are all capitalized now! :)

**Pull request merged**